File: python-backend/jar-runner/stubs/com/github/catvod/crawler/Spider.java
package com.github.catvod.crawler;

import android.content.Context;

/**
 * TVBox Spider 基类 stubs（兼容原始 jar 中继承此基类的 spider）。
 * 仅提供方法签名，方法体为空（由子类覆盖）。
 */
public class Spider {
    public void init(Context context, String extend) throws Exception {}

    public void init(Context context) throws Exception { init(context, ""); }

    public String homeContent(boolean filter) throws Exception { return ""; }

    public String homeVideoContent() throws Exception { return ""; }

    public String categoryContent(String tid, String pg, boolean filter, String extend) throws Exception { return ""; }

    /** 部分 jar 蜘蛛以 HashMap 形式传递 extend（与 String 版签名并存）。 */
    public String categoryContent(String tid, String pg, boolean filter, java.util.HashMap<String, String> extend)
            throws Exception {
        return "";
    }

    public String detailContent(String[] ids) throws Exception { return ""; }

    public String searchContent(String key, boolean quick, String pg) throws Exception { return ""; }

    public String playerContent(String flag, String id, String[] vipFlags) throws Exception { return ""; }

    public String liveContent(String url) throws Exception { return ""; }

    public String proxy(String param) throws Exception { return ""; }

    public boolean isVideoFormat(String url) { return false; }

    public boolean manualVideoCheck() { return false; }

    public void destroy() {}

    /**
     * 真实 CatVod SDK 的共享 OkHttpClient（jar 内 merge/AC/G.E() 优先调用）。
     * 带 20s 连接/读写超时：慢站点快速失败，避免蜘蛛网络请求无限挂起
     * 导致整个 JVM 桥超时重启。（编译依赖 vendor/dexdeps/okhttp3.jar）
     */
    public static okhttp3.OkHttpClient client() {
        return new okhttp3.OkHttpClient.Builder()
                .connectTimeout(20, java.util.concurrent.TimeUnit.SECONDS)
                .readTimeout(20, java.util.concurrent.TimeUnit.SECONDS)
                .writeTimeout(20, java.util.concurrent.TimeUnit.SECONDS)
                .build();
    }

    /** 部分 jar 蜘蛛调用 Spider.safeDns() 获取 DNS；系统默认即可。 */
    public static okhttp3.Dns safeDns() {
        return okhttp3.Dns.SYSTEM;
    }
}

File: python-backend/base/spider.py
import re
import os
import json
import time
import urllib.parse
import hoststate
import http_client
from lxml import etree
from abc import abstractmethod, ABCMeta
from importlib.machinery import SourceFileLoader


class Spider(metaclass=ABCMeta):
    _instance = None

    def __init__(self):
        self.extend = ''

    def __new__(cls, *args, **kwargs):
        if cls._instance:
            return cls._instance
        cls._instance = super().__new__(cls)
        return cls._instance

    @abstractmethod
    def init(self, extend=''):
        pass

    def homeContent(self, filter):
        pass

    def homeVideoContent(self):
        pass

    def categoryContent(self, tid, pg, filter, extend):
        pass

    def detailContent(self, ids):
        pass

    def searchContent(self, key, quick, pg='1'):
        pass

    def playerContent(self, flag, id, vipFlags):
        pass

    def liveContent(self, url):
        pass

    def localProxy(self, param):
        pass

    def isVideoFormat(self, url):
        pass

    def manualVideoCheck(self):
        pass

    def action(self, action):
        pass

    def destroy(self):
        pass

    def getName(self):
        pass

    def getDependence(self):
        return []

    def loadSpider(self, name):
        return self.loadModule(name).Spider()

    def loadModule(self, name):
        cache_dir = hoststate.get_cache_dir()
        path = os.path.join(os.path.join(cache_dir, 'py'), f'{name}.py')
        return SourceFileLoader(name, path).load_module()

    def regStr(self, reg, src, group=1):
        m = re.search(reg, src)
        src = ''
        if m:
            src = m.group(group)
        return src

    def removeHtmlTags(self, src):
        clean = re.compile('<.*?>')
        return re.sub(clean, '', src)

    def cleanText(self, src):
        clean = re.sub('[😀-🙏🌀-🗿🚀-\U0001f6ff\U0001f1e0-🇿]', '', src)
        return clean

    def fetch(self, url, params=None, cookies=None, headers=None, timeout=5, verify=True, stream=False, allow_redirects=True):
        rsp = http_client.get(url, params=params, cookies=cookies, headers=headers, timeout=timeout, verify=verify, stream=stream, allow_redirects=allow_redirects)
        rsp.encoding = 'utf-8'
        return rsp

    def post(self, url, params=None, data=None, json=None, cookies=None, headers=None, timeout=5, verify=True, stream=False, allow_redirects=True):
        rsp = http_client.post(url, params=params, data=data, json=json, cookies=cookies, headers=headers, timeout=timeout, verify=verify, stream=stream, allow_redirects=allow_redirects)
        rsp.encoding = 'utf-8'
        return rsp

    def html(self, content):
        return etree.HTML(content)

    @staticmethod
    def str2json(text):
        return json.loads(text)

    @staticmethod
    def json2str(obj):
        return json.dumps(obj, ensure_ascii=False)

    def getProxyUrl(self, local=True):
        return f"{hoststate.get_proxy_url(local)}?do=py"

    def log(self, msg):
        if isinstance(msg, dict) or isinstance(msg, list):
            print(json.dumps(msg, ensure_ascii=False))
        else:
            print(f'{msg}')

    def getCache(self, key):
        # key 需 URL 编码（M-22）：含 &/=/#/中文等字符的 key 不编码会被
        # 拆成多个查询参数或截断
        k = urllib.parse.quote(str(key), safe='')
        value = self.fetch(f'http://127.0.0.1:{hoststate.get_port()}/cache?do=get&key={k}', timeout=5).text
        if len(value) > 0:
            if value.startswith('{') and value.endswith('}') or value.startswith('[') and value.endswith(']'):
                value = json.loads(value)
                if type(value) == dict:
                    if 'expiresAt' not in value or value['expiresAt'] >= int(time.time()):
                        return value
                    else:
                        self.delCache(key)
                        return None
            return value
        else:
            return None

    def setCache(self, key, value):
        if type(value) in (int, float):
            value = str(value)
        if len(value) > 0:
            if type(value) == dict or type(value) == list:
                value = json.dumps(value, ensure_ascii=False)
        k = urllib.parse.quote(str(key), safe='')
        r = self.post(f'http://127.0.0.1:{hoststate.get_port()}/cache?do=set&key={k}', data={'value': value}, timeout=5)
        return 'succeed' if r.status_code == 200 else 'failed'

    def delCache(self, key):
        k = urllib.parse.quote(str(key), safe='')
        r = self.fetch(f'http://127.0.0.1:{hoststate.get_port()}/cache?do=del&key={k}', timeout=5)
        return 'succeed' if r.status_code == 200 else 'failed'
